overlord/snapstate/backend/copydata.go

```diff
@@ -86,16 +92,137 @@ func (b Backend) UndoCopySnapData(newInfo, oldInfo *snap.Info, _ progress.Meter,
 }
 
 // ClearTrashedData removes the trash. It returns no errors on the assumption that it is called very late in the game.
-func (b Backend) ClearTrashedData(oldSnap *snap.Info, opts *dirs.SnapDirOptions) {
-	dirs, err := snapDataDirs(oldSnap, opts)
+func (b Backend) ClearTrashedData(oldSnap *snap.Info) {
+	dataDirs, err := snapDataDirs(oldSnap, nil)
 	if err != nil {
 		logger.Noticef("Cannot remove previous data for %q: %v", oldSnap.InstanceName(), err)
 		return
 	}
 
-	for _, d := range dirs {
+	opts := &dirs.SnapDirOptions{HiddenSnapDataDir: true}
+	hiddenDirs, err := snapDataDirs(oldSnap, opts)
+	if err != nil {
+		logger.Noticef("Cannot remove previous data for %q: %v", oldSnap.InstanceName(), err)
+		return
+	}
+
+	// this will have duplicates but the second remove will just be ignored
+	dataDirs = append(dataDirs, hiddenDirs...)
+	for _, d := range dataDirs {
 		if err := clearTrash(d); err != nil {
 			logger.Noticef("Cannot remove %s: %v", d, err)
 		}
 	}
 }
+
+func (b Backend) HideSnapData(snapName string) error {
+	postMigrationOpts := &dirs.SnapDirOptions{HiddenSnapDataDir: true}
+
+	users, err := allUsers(nil)
+	if err != nil {
+		return err
+	}
+
+	for _, usr := range users {
+		uid, gid, err := osutil.UidGid(usr)
+		if err != nil {
+			return err
+		}
+
+		// nothing to migrate
+		oldSnapDir := snap.UserSnapDir(usr.HomeDir, snapName, nil)
+		if _, err := os.Stat(oldSnapDir); errors.Is(err, os.ErrNotExist) {
+			continue
+		} else if err != nil {
+			return fmt.Errorf("cannot stat snap dir %q: %w", oldSnapDir, err)
```

Comment:
I wonder if this shouldn't be a `continue` as well. Since we are running as root we don't need to worry about permissions, but `stat` can also fail if the level of symbolic links it too deep, and this is something that a malicious user can control.

I cannot see a real security problem here, and probably just continuing here can also cause troubles (if the error was `ENOMEM`, for example), but maybe adding more errors (such as ELOOP, but possibly also ENAMETOOLONG and ENOTDIR) to the first branch (the "continue" one) might improve robustness.

Reply:
The issue is that continuing here would leave us in a state where we're partially migrated (some users are and some aren't). I guess the user could fix the underlying issue, disable the feature and then re-enable it but since the feature is tied to refreshes it's not as easy